### PR TITLE
vidconv: add support for converting to/from YUV444 and RGB32

### PR DESCRIPTION
**Pull request by alfredh**
Add support for video converting:

- From YUV444 to RGB32
- From RGB32 to YUV444

tested with Baresip and Retest.

This is the last patch for the YUV444 pixelformat for now :)

### CODE REVIEW

File: src/vidconv/vconv.c [hunk 1]
```diff
@@ -180,6 +180,21 @@ static inline void yuv2rgb555(uint8_t *rgb, uint8_t y,
 }
 
 
+static inline void _yuv2rgb(uint8_t *rgb, uint8_t y, uint8_t u, uint8_t v)
```
Comment (richaas) [inline src/vidconv/vconv.c:183]: we already have a yuv2rgb() function, similar to this. Let's reuse that instead.

File: src/vidconv/vconv.c [hunk 2]
```diff
@@ -342,6 +357,57 @@ static void rgb32_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+			     const uint8_t *ds2, unsigned lss
+			     )
+{
+	unsigned x, xd, xs, xs2;
+	unsigned id;
+
+	(void)ds1;
+	(void)ds2;
+
+	for (x=0; x<width; x+=2) {
```
Comment (richaas) [inline src/vidconv/vconv.c]: I think we can simplify by doing x++ instead of x+=2.

File: src/vidconv/vconv.c [hunk 3]
```diff
@@ -573,6 +639,42 @@ static void nv21_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+			     const uint8_t *ds2, unsigned lss)
+{
+	unsigned x, xd, xs, xs2;
+	unsigned id;
+	unsigned is1, is2, is3, is4;
+
+	(void)dd1;
+	(void)dd2;
+
+	for (x=0; x<width; x+=2) {
```
Comment (richaas) [inline src/vidconv/vconv.c]: Same as above, I think x++ is better here.

File: src/vidconv/vconv.c (hunk 1 above)
Comment (alfredh) [reply src/vidconv/vconv.c:183]: for pixelformats such as YUV420P we calculate a common color for the whole 2x2 block.
for YUV444 each pixel has its own color.

yuv2rgb is used to convert translated YUV values (int ruv, int guv, int buv)
but _yuv2rgb takes raw YUV values as input.

**Comment by alfredh** [comment]
Thanks @richaas for review. I updated the PR with some improvements.

File: src/vidconv/vconv.c [hunk 4]
```diff
@@ -342,6 +354,46 @@ static void rgb32_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+	unsigned x, xd, xs;
+	unsigned id;
+
+	(void)ds1;
+	(void)ds2;
+
+	for (x=0; x<width; x++) {
+
+		uint32_t x0;
+		uint32_t x2;
```
Comment (richaas) [inline src/vidconv/vconv.c]: x1?

File: src/vidconv/vconv.c [hunk 5]
```diff
@@ -342,6 +354,46 @@ static void rgb32_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+
+		xd  = x + xoffs;
+
+		xs  = 4 * ((unsigned)( x * rw));
```
Comment (richaas) [inline src/vidconv/vconv.c]: whitespace

File: src/vidconv/vconv.c [hunk 6]
```diff
@@ -342,6 +354,46 @@ static void rgb32_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+
+	(void)ds1;
+	(void)ds2;
+
+	for (x=0; x<width; x++) {
+
+		uint32_t x0;
+		uint32_t x2;
+
+		xd  = x + xoffs;
```
Comment (richaas) [inline src/vidconv/vconv.c]: whitespace

File: src/vidconv/vconv.c [hunk 7]
```diff
@@ -342,6 +354,46 @@ static void rgb32_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+
+		xs  = 4 * ((unsigned)( x * rw));
+
+		id = xd + yd*lsd;
+
+		x0 = *(uint32_t *)(void *)&ds0[xs  + ys *lss];
+		x2 = *(uint32_t *)(void *)&ds0[xs  + ys2*lss];
```
Comment (richaas) [inline src/vidconv/vconv.c]: whitespace x 2

File: src/vidconv/vconv.c [hunk 8]
```diff
@@ -342,6 +354,46 @@ static void rgb32_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+
+		dd0[id]         = rgb2y(x0 >> 16, x0 >> 8, x0);
```
Comment (richaas) [inline src/vidconv/vconv.c]: whitespace x 6

File: src/vidconv/vconv.c [hunk 9]
```diff
@@ -573,6 +625,37 @@ static void nv21_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+	unsigned x, xd, xs;
+	unsigned id;
+	unsigned is1, is2;
+
+	(void)dd1;
+	(void)dd2;
+
+	for (x=0; x<width; x++) {
+
+		xd  = (x + xoffs) * 4;
```
Comment (richaas) [inline src/vidconv/vconv.c]: whitespace

File: src/vidconv/vconv.c [hunk 10]
```diff
@@ -573,6 +625,37 @@ static void nv21_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+
+		xs  = (unsigned)(x * rw);
```
Comment (richaas) [inline src/vidconv/vconv.c]: whitespace

File: src/vidconv/vconv.c [hunk 11]
```diff
@@ -573,6 +625,37 @@ static void nv21_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+
+		id = (xd + yd*lsd);
+
+		is1  = (xs) + (ys )*lss;
```
Comment (richaas) [inline src/vidconv/vconv.c]: whitespace x 2

File: src/vidconv/vconv.c [hunk 12]
```diff
@@ -573,6 +625,37 @@ static void nv21_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+		is2  = (xs) + (ys2)*lss;
+
+		_yuv2rgb(&dd0[id],       ds0[is1], ds1[is1], ds2[is1] );
```
Comment (richaas) [inline src/vidconv/vconv.c]: whitespace x 2

File: src/vidconv/vconv.c [hunk 13]
```diff
@@ -573,6 +625,37 @@ static void nv21_to_yuv420p(unsigned xoffs, unsigned width, double rw,
...
+	(void)dd1;
+	(void)dd2;
+
+	for (x=0; x<width; x++) {
+
+		xd = (x + xoffs) * 4;
+
+		xs = (unsigned)(x * rw);
+
+		id = (xd + yd*lsd);
```
Comment (richaas) [inline src/vidconv/vconv.c]: id, is1 and is2 calculation: remove redundant parentheses.

**Pull request merged**